### PR TITLE
Upgrade the fabric client to support newer versions of k8s

### PR DESCRIPTION
**Pull request by churromorales**
This PR contains 2 items. 

1. Upgrade the k8s client to support newer versions of k8s. 
2. Ability to override specific Metrics Monitors 
 
We found that when we used the `TaskCountStatsMonitor` in the overlord config, the peon tasks would not start because they inherit the monitors from the parent process, which used to be the Middle Manager (but that would never have that monitor originally).  So now you can override this value with the following config: 

`druid.indexer.runner.peonMonitors`

### CODE REVIEW

File: docs/development/extensions-contrib/k8s-jobs.md
Comment (a2l007) [inline docs/development/extensions-contrib/k8s-jobs.md:69]: Nit: It would be useful to mention this property in the [extension specific docs](https://druid.apache.org/docs/latest/development/extensions-contrib/k8s-jobs.html) as well.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/LogWatchInputStream.java
Comment (a2l007) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/LogWatchInputStream.java:28]: Can we add javadocs for this class?

Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/LogWatchInputStream.java:28]: sure

Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/LogWatchInputStream.java:33]: I believe the issue with testing is invalid.  I just upgraded a client library.  I did have to change code but I did not change method contracts or behavior so existing tests should all pass.  I don’t think I can add any meangful tests here without just eating up cpu cycles.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java
Comment (a2l007) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java:108]: Why is this change needed?

Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java:108]: This was a tricky one.  For some background 

the overlord launches a task, waits for it to complete and then returns the status.  Now in k8s < 1.25 it would just mean once the pod is not active, go grab the status.
Now the contact has changed a bit, because in 1.25 they introduced finalizers:
https://github.com/kubernetes/kubernetes/pull/110948

And we noticed that when we ran tasks on a 1.25 k8s  druid cluster they would complete fine but marked as failure.
We outputted the job status and noticed that the job was not active, but both success and failure were not set, but there was this field that had
```
uncountedTerminatedPods=UncountedTerminatedPods(failed=[], succeeded=[e916cbf9-467a-45f3-86a7-3767145d6384], additionalProperties={})
```
which from the docs:
```
UncountedTerminatedPods holds the UIDs of Pods that have terminated but the job controller hasn't yet accounted for in the status counters. The job controller creates pods with a finalizer. When a pod terminates (succeeded or failed), the controller does three steps to account for it in the job status: (1) Add the pod UID to the arrays in this field. (2) Remove the pod finalizer. (3) Remove the pod UID from the arrays while increasing the corresponding counter. This field is beta-level. The job controller only makes use of this field when the feature gate JobTrackingWithFinalizers is enabled (enabled by default). Old jobs might not be tracked using this field, in which case the field remains null.
```
So now what happens is the job goes from a state where it is not active, to having uncountedTerminatedPods to then having a status with success or failure.  I will push up a one-line fix to make this work, but for those of you working with 1.25 version of k8s, I’m sure you will be affected as well.

Basically add another check to wait on,
Right now we wait for this:
```
// block until
job.getStatus() != null && job.getActive() == null
// then return 
return job.getStatus().getSucceeded() != null
```
So the change has to become
```
// block until
job.getStatus() != null && job.getActive() == null && (job.getStatus().getFailed() != null || job.getStatus().getSucceeded() !=null)
// then return 
return job.getStatus().getSucceeded() != null
```
This should keep things backwards compatible and working in all versions of k8s

**Review by a2l007 — APPROVED** [review]

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java
Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java:108]: @abhishekagarwal87 what do you think?  This should bring the client up-to-date as well as fix things for those druid users running on k8s 1.25+ that were having issues.

**Comment by churromorales** [comment]
@abhishekagarwal87 what do you think?  This should bring the client up-to-date as well as fix things for those druid users running on k8s 1.25+ that were having issues.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java
Comment (abhishekagarwal87) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java:259]: this log line doesn't really include any task-specific info. can you add that too?

File: extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidKubernetesClientTest.java
Comment (abhishekagarwal87) [inline extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidKubernetesClientTest.java:37]: Hmm. This is a bit of a bummer. Is there no way to avoid this?

File: extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidPeonClientIntegrationTest.java
Comment (abhishekagarwal87) [inline extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidPeonClientIntegrationTest.java:57]: We are using K3S now. do you still same issues?

File: extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/K8sTaskAdapterTest.java
Comment (abhishekagarwal87) [inline extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/K8sTaskAdapterTest.java:227]: Please print the container.getEnv if the assertion fails.

File: docs/development/extensions-contrib/k8s-jobs.md
Comment (ektravel) [inline docs/development/extensions-contrib/k8s-jobs.md:69]:
```suggestion
|`druid.indexer.runner.peonMonitors`|`JsonArray`|Overrides `druid.monitoring.monitors`. Use this property if you don't want to inherit monitors from the overlord.|`[]`|No|
```

**Review by ektravel — COMMENTED** [review]
Left one suggestion for the doc.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java
Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java:108]: that could work, want me to change it?  This definitely works above, I don't have access to many different k8s environments, and someone in the community helped me test the above on 1.25.  I could definitely change this, but I unfortunately have no way of testing this out.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java
Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java:129]: ```List<StatusDetails> delete();```  used to return true or false i believe.

File: extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidPeonClientIntegrationTest.java
Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidPeonClientIntegrationTest.java:57]: k3s?  This was for anyone who had an existing k8s cluster, they could just run this test and point it there.  The k8s integration tests are definitely not flexible enough for me to test mm-less without rewriting a lot of it.

File: extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidKubernetesClientTest.java
Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidKubernetesClientTest.java:37]: So I had tests for a class, then i upgraded a client library and had to change some things.  but the functions were all tested before.  Now i would expect that if tests pass I would be good, since it was already tested before.  But turns out if you have to make any changes to make the new library apis work, then code coverage complains.  It is unfortunate here, because I didn't change any behavior.

File: extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/K8sTaskAdapterTest.java
Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/K8sTaskAdapterTest.java:227]: sure

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java
Comment (nlippis) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java:108]: Please make this change, we can test it out on a v1.25 cluster

**Comment by abhishekagarwal87** [comment]
@churromorales - can you address the build failures?

File: docs/development/extensions-contrib/k8s-jobs.md
Comment (techdocsmith) [inline docs/development/extensions-contrib/k8s-jobs.md:69]:
```suggestion
|`druid.indexer.runner.peonMonitors`|`JsonArray`|Overrides `druid.monitoring.monitors`. Use this property if you don't want to inherit monitors from the Overlord.|`[]`|No|
```
nit. capitalize Druid service names.

**Review by techdocsmith — CHANGES_REQUESTED** [review]
Small stylistic change for doc.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java
Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/DruidKubernetesPeonClient.java:108]: @nlippis I believe you can test by taking this commit and making that small change and apply it onto the 25 branch.  I think if it works for you, ill resubmit this with the change.

**Comment by churromorales** [comment]
> @churromorales - can you address the build failures?
I addressed them in a comment above, but maybe it was missed. 

So I had tests for the `DruidKubernetesPeonClient` class, then i upgraded a client library and had to change some things. but the functions were all tested before. Now i would expect that if tests pass I would be good, since it was already tested before. But turns out if you have to make any changes to make the new library apis work, then code coverage complains. It is unfortunate here, because I didn't change any behavior.  I can add more tests, but they wont be meaningful here, I believe the build problems are all because I updated the library and had to make some code changes to make it work.  There are no logic changes here, except for the issue in 1.25, which I did test.

**Review by jwitko — APPROVED** [review]
very much needed changes.  thank you for making

File: extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidKubernetesClientTest.java
Comment (abhishekagarwal87) [reply extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidKubernetesClientTest.java:37]: aren't there any tests covering that path?

File: extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidPeonClientIntegrationTest.java
Comment (abhishekagarwal87) [reply extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidPeonClientIntegrationTest.java:57]: how was this running before?

**Comment by churromorales** [comment]
@abhishekagarwal87 you good with this now?  I explained about the test coverage in an earlier comment.

File: extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidPeonClientIntegrationTest.java
Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidPeonClientIntegrationTest.java:57]: I had a kind cluster running locally, it was in the original patch.  For anyone that wants to test with a local k8s cluster I left this test in the codebase.  If you want me to remove it, I'll be happy to get rid of it.  I find it useful for when I make changes, since I can't integration test this with the way current druid-it tests are setup.

**Comment by churromorales** [comment]
> Do you also need to make this change?
> 
> ```diff
> --- a/extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java
> +++ b/extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java
> @@ -180,7 +180,7 @@ public class KubernetesTaskRunner implements TaskLogStreamer, TaskRunner
>                      completedPhase = monitorJob(peonPod, k8sTaskId);
>                    } else {
>                      Job job = existingJob.get();
> -                    if (job.getStatus().getActive() == null) {
> +                    if (job.getStatus() != null && job.getStatus().getActive() == null && (job.getStatus().getFailed() != null || job.getStatus().getSucceeded() !=null)) {
>                        if (job.getStatus().getSucceeded() != null) {
>                          completedPhase = new JobResponse(job, PeonPhase.SUCCEEDED);
>                        } else {
> ```

I believe that is the crux of the change.  I explain why here: https://github.com/apache/druid/pull/13804#discussion_r1112268989  Or let me know if I misunderstood you.

File: extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidKubernetesClientTest.java
Comment (churromorales) [reply extensions-contrib/kubernetes-overlord-extensions/src/test/java/org/apache/druid/k8s/overlord/common/DruidKubernetesClientTest.java:37]: i added a getter to a class, the code coverage tool complained that i added new code without a test.  So I added a test for the getter.  That was new code, so i guess its okay to add a test if that's what druid wants.  Although I feel tests like this are not really that useful and make the whole suite slow in general.

**Comment by a2l007** [comment]
@abhishekagarwal87 Are there any additional concerns with this PR apart from the code coverage failure? From the discussion it looks like there isn't really a way to make the coverage tool pass without adding non-meaningful tests.

**Comment by mindreader** [comment]
Or let me know if I misunderstood you.




> > Do you also need to make this change?
> > ```diff
> > --- a/extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java
> > +++ b/extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java
> > @@ -180,7 +180,7 @@ public class KubernetesTaskRunner implements TaskLogStreamer, TaskRunner
> >                      completedPhase = monitorJob(peonPod, k8sTaskId);
> >                    } else {
> >                      Job job = existingJob.get();
> > -                    if (job.getStatus().getActive() == null) {
> > +                    if (job.getStatus() != null && job.getStatus().getActive() == null && (job.getStatus().getFailed() != null || job.getStatus().getSucceeded() !=null)) {
> >                        if (job.getStatus().getSucceeded() != null) {
> >                          completedPhase = new JobResponse(job, PeonPhase.SUCCEEDED);
> >                        } else {
> > ```
> 
> I believe that is the crux of the change. I explain why here: [#13804 (comment)](https://github.com/apache/druid/pull/13804#discussion_r1112268989) Or let me know if I misunderstood you.

Sorry I wasn't clear. There are two spots in the code that have that same success / failure logic, I only see one in this pr, after the job is monitored. But I also see one before the job is monitored just after it is spawned in KubernetesTaskRunner. If the logic is not necessary at that point, please feel free to ignore me.

**Comment by abhishekagarwal87** [comment]
@a2l007 - Sorry if I am missing something but if CC is failing anyway, there is no need to add `makingCodeCoverageHappy()` test in this PR. That's the only comment I have. LGTM otherwise.

**Comment by churromorales** [comment]
> 

it is a good question.  I think the way I coded it was not very clear, I apologize for that. 
I tried to make the `KubernetesTaskRunnerTest` cover the different scenarios.  

Lets look at this case....

startup overlord, 
1st thing it does is go to k8s and see if there are any running tasks.  If there are tasks, watch them and do bookkeeping when they finish. 

The conditional block you are referring to handles that case.   It first does a check to see if the task is not active, then checks the state.  If it is still running it does the `if` statement you suggested (in following lines to monitorJob())`, so no need to do it twice.    It's just a short circuit in case its already finished, grab the status and return, no need to make more k8s calls.

Does that help make sense of how things are working?  These are good questions, and I want to make sure the code is doing the correct thing.

**Comment by churromorales** [comment]
@abhishekagarwal87 The test failing are from the other PR that was merged, I don't have cycles to look and fix whatever is wrong and we honestly forked our druid so whether it goes upstream is not a big deal.  

I think the thing is the other PR is using all the old libraries, and it can't deserialize the pod template anymore.  So if you want to update this feature to work with newer k8s versions, that will have to be sorted out at some point, or you can just close off this PR and worry about upgrades when the time comes.  As it stands the mm-less wont work in newer versions of k8s so, I'll leave it up to imply what they want to do.

**Comment by abhishekagarwal87** [comment]
I understand @churromorales. would you be open to giving write access on your branch to someone who wants to take this PR forward?

**Comment by clintropolis** [comment]
closing since these changes were added to #14028

**Pull request closed without merging**
